flash_io.c

```diff
@@ -0,0 +1,413 @@
+#include <hardware/flash.h>
+#include <hardware/sync.h>
+#include <pico/time.h>
+
+#include <string.h>
+
+#include "common.h"
+#include "flash_io.h"
+
+#define FLASH_PAGE_COUNT_IN_SECTOR FLASH_SECTOR_SIZE / FLASH_PAGE_SIZE
+#define FLASH_MAX_PAGE_INDEX FLASH_PAGE_COUNT_IN_SECTOR - 1
+
+// Identifier to distinguish between random bytes and our data in flash memory.
+#define TAG 0xAAAAAAAAAAAAAAAA
+
+// this symbol is defined in the linker script (memmap.ld)
+extern size_t
+FLASH_DEVICE_DATA_START;
+
+enum __attribute__((packed))
+init_targets {
+  MAGNETRON = 1 << 0,
+  WATER     = 1 << 1,
+  PILOT     = 1 << 2,
+  MAPPER    = 1 << 3,
+  SHUTTER   = 1 << 4,
+  STIRER    = 1 << 5,
+
+  NONE      = 1 << 7      /*
+                            * NONE must be last.
+                            * This is done to distinguish "-1" from valid init_targets.
+                            * By default, flash memory is set to "-1".
+                            */
+};
```

Comment:
The algorithm does not expect init_targets to be 1 byte in size. However, as described in the code above, we need to distinguish between -1 and a valid target. We do expect init_targets to be 1 byte in size because NONE is set to 1 << 7.

That said, to ensure that NONE is last, I'm adding 
`_Static_assert(NONE == 1 << (sizeof(enum init_target) * 8) - 1)`